From dd710281580135b40c98e95bc6a9ad562052f010 Mon Sep 17 00:00:00 2001
From: Dudemanguy <random342@airmail.cc>
Date: Thu, 22 Jan 2026 11:12:22 -0600
Subject: [PATCH 1/2] opengl/context: require swap_buffers param for FenceSync

Commits 8854c742711705be1fcfcc6a5875960a3e2593dc and
5ae0e0ff9a6974453c1aba41ddc8442d12ac9264 eliminated the external
swapchain API that the render backend (vo_libmpv) was previously using.
However, it was missed that this was being used to skip the
gl->FenceSync calls and that logic change was not properly accounted
for. This meant that vo_libmpv would leak fences since
ra_gl_ctx_swap_buffers is never called to clean them up.

Fix this by making sure the ra_ctx_params have swap_buffers defined
before calling gl->FenceSync. All platforms are required to implement
this anyway with the exception of vo_libmpv since it's a funny special
case. Fixes #17217.
---
 video/out/opengl/context.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/video/out/opengl/context.c b/video/out/opengl/context.c
index 86e8d5daac8b4..86e5c797d670a 100644
--- a/video/out/opengl/context.c
+++ b/video/out/opengl/context.c
@@ -232,7 +232,7 @@ bool ra_gl_ctx_submit_frame(struct ra_swapchain *sw, const struct vo_frame *fram
     if (p->opts->use_glfinish)
         gl->Finish();
 
-    if (gl->FenceSync) {
+    if (gl->FenceSync && p->params.swap_buffers) {
         GLsync fence = gl->FenceSync(GL_SYNC_GPU_COMMANDS_COMPLETE, 0);
         if (fence)
             MP_TARRAY_APPEND(p, p->vsync_fences, p->num_vsync_fences, fence);

From 2eb4c8c38cb3e5213bd94e29fc50c2354f2f22db Mon Sep 17 00:00:00 2001
From: Dudemanguy <random342@airmail.cc>
Date: Fri, 23 Jan 2026 09:48:18 -0600
Subject: [PATCH 2/2] opengl/context: remove unneeded ra_swapchain_fns
 redirection

We don't allow backends to override this anymore since
5ae0e0ff9a6974453c1aba41ddc8442d12ac9264 so we can just set it directly
in ra_swapchain.
---
 video/out/opengl/context.c | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/video/out/opengl/context.c b/video/out/opengl/context.c
index 86e5c797d670a..31675dc4f23e8 100644
--- a/video/out/opengl/context.c
+++ b/video/out/opengl/context.c
@@ -79,7 +79,6 @@ struct priv {
     struct mp_log *log;
     struct ra_ctx_params params;
     struct opengl_opts *opts;
-    struct ra_swapchain_fns fns;
     GLuint main_fb;
     struct ra_tex *wrapped_fb; // corresponds to main_fb
     // for debugging:
@@ -130,6 +129,7 @@ bool ra_gl_ctx_init(struct ra_ctx *ctx, GL *gl, struct ra_ctx_params params)
     struct ra_swapchain *sw = ctx->swapchain = talloc_ptrtype(NULL, sw);
     *sw = (struct ra_swapchain) {
         .ctx = ctx,
+        .fns = &ra_gl_swapchain_fns,
     };
 
     struct priv *p = sw->priv = talloc_ptrtype(sw, p);
@@ -138,11 +138,8 @@ bool ra_gl_ctx_init(struct ra_ctx *ctx, GL *gl, struct ra_ctx_params params)
         .log    = ctx->log,
         .params = params,
         .opts   = mp_get_config_group(p, ctx->global, &opengl_conf),
-        .fns    = ra_gl_swapchain_fns,
     };
 
-    sw->fns = &p->fns;
-
     if (!gl->version && !gl->es)
         return false;
 
